Implement Azure ResourceTags

fahlmant opened the pull request:
**What this PR does / why we need it**:
Implements ResourceTags for Azure to match the equivalent AWS capabilities. Specifically, tags to be set at the HostedCluster Platform level, as well as on individual NodePools. 

## Which issue(s) this PR fixes:
No Jira

## Special notes for your reviewer:

N/A 

**Checklist**
- [X] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [x] This change includes docs. 
- [x] This change includes unit tests.

bryan-cox commented:
Please convert this to a draft this to a draft PR until its ready for review so it doesn't trigger CI jobs on every push.

jparrill commented:
The `Verify` test is failing due to:
```
Linting commits from 6b857ec154a004563e0bf4df4493ecd646c32917 to 2d0334cc8d0741333d14352f3f1fa23e03c0940c (CI: PR targeting 6b857ec154a004563e0bf4df4493ecd646c32917)
1: CT1 Title does not follow ConventionalCommits.org format 'type(optional-scope): description': "Implement Azure ResourceTags"
3: B6 Body message is missing
```

bryan-cox commented on api/hypershift/v1beta1/azure.go:
```diff
@@ -98,6 +98,29 @@ type AzureNodePoolPlatform struct {
+	// +listType=map
+	// +listMapKey=key
+	// +optional
+	ResourceTags []AzureResourceTag `json:"resourceTags,omitempty"`
```
Will any changes to this field cause the a rollout of new nodes? If so that should be documented.

`// +rollout` should also be added here if so.

bryan-cox commented on hypershift-operator/controllers/nodepool/azure.go:
```diff
@@ -95,11 +95,42 @@ func azureMachineTemplateSpec(nodePool *hyperv1.NodePool) (*capiazure.AzureMachi
 }
 
+// azureAdditionalTags merges tags from nodepool and cluster, with cluster tags taking precedence
+func azureAdditionalTags(hostedCluster *hyperv1.HostedCluster, nodePool *hyperv1.NodePool) map[string]string {
```
Please add a unit test for this function.

bryan-cox commented on hypershift-operator/controllers/nodepool/azure.go:
```diff
@@ -95,11 +95,42 @@ func azureMachineTemplateSpec(nodePool *hyperv1.NodePool) (*capiazure.AzureMachi
 }
 
+// azureAdditionalTags merges tags from nodepool and cluster, with cluster tags taking precedence
+func azureAdditionalTags(hostedCluster *hyperv1.HostedCluster, nodePool *hyperv1.NodePool) map[string]string {
```
This automatically allocates a map even when there are no tags. There are efficiency gains when allocating the map to the same size as the number of tags like

```
func azureAdditionalTags(hostedCluster *hyperv1.HostedCluster, nodePool *hyperv1.NodePool) map[string]string {
	var nodePoolTags, clusterTags []hyperv1.AzureResourceTag

	// Extract tag slices
	if nodePool.Spec.Platform.Azure != nil {
		nodePoolTags = nodePool.Spec.Platform.Azure.ResourceTags
	}
	if hostedCluster.Spec.Platform.Azure != nil {
		clusterTags = hostedCluster.Spec.Platform.Azure.ResourceTags
	}

	totalSize := len(nodePoolTags) + len(clusterTags)
	if totalSize == 0 {
		return nil
	}

	// Pre-allocate with known capacity for optimal performance
	// This eliminates map resize operations and reduces memory allocations by ~73%
	tags := make(map[string]string, totalSize)

	// Add nodepool tags first
	for _, tag := range nodePoolTags {
		tags[tag.Key] = tag.Value
	}

	// Add cluster tags (which take precedence over nodepool tags)
	for _, tag := range clusterTags {
		tags[tag.Key] = tag.Value
	}

	return tags
}
```

I used a Claude golang agent to run a benchmark. This is the results of the new function vs the old:
```
Key Results:

  - Empty tags: 4.7x faster (7.69ns → 1.62ns)
  - Many tags (50): 1.7x faster with 42% less memory usage
  - Memory allocations: Reduced from 7 to 4 allocations (43% fewer)

  The Main Optimizations:

  1. Early return optimization:
  // Old: Always allocates map, checks later
  tags := make(map[string]string)
  // ... add tags ...
  if len(tags) == 0 { return nil }

  // New: Check first, allocate only if needed
  if totalSize == 0 { return nil }
  tags := make(map[string]string, totalSize)
  2. Pre-allocation with capacity:
  // Old: Map grows dynamically (expensive resizes)
  tags := make(map[string]string)

  // New: Single allocation with exact size
  tags := make(map[string]string, totalSize)

  Real-World Impact:

  - Empty tag scenarios (common): 4.7x faster
  - High tag count scenarios: 1.7x faster + 42% less memory
  - Reduced GC pressure: 43% fewer allocations
```

fahlmant replied:
Agreed and changed

bryan-cox reviewed (commented):
Thanks for splitting this up in commits! That was really helpful.

I'm going to remove the NO-JIRA from the title of this PR. This change is large enough that we should have a Jira story behind this change documenting things like: who will test this change, why do we need this change, who is requesting this change, what do we gain from the change, etc.

fahlmant committed:
feat(azure): add resource tags

Add ResourceTags to Azure platform and NodePool API specs. This enables
Azure resources to be tagged with a uniform set of tags, matching the
equivalent AWS Platform and NodePool ResourceTags fields.

fahlmant committed:
feat(azure): add generated files from resource tags change

Results of running make api after Azure ResourceTags API changes.

fahlmant committed:
feat(azure): Implement resource tag handling in NodePool controller

In the Azure implementation of the NodePool controller, add
functionality to apply the tags provided in the ResourceTag spec
to the Azure MachineTemplate.

fahlmant commented on api/hypershift/v1beta1/azure.go:
```diff
@@ -98,6 +98,29 @@ type AzureNodePoolPlatform struct {
 	Diagnostics *Diagnostics `json:"diagnostics,omitempty"`
+
+	// resourceTags is an optional list of additional tags to apply to Azure node
+	// instances. Changes to this field will be propagated in-place to Azure VM instances and their initial OS disks.
```
Removed as this is not the case

fahlmant commented on api/hypershift/v1beta1/azure.go:
```diff
@@ -735,3 +780,24 @@ type AzureAuthenticationConfiguration struct {
+	// key is the key part of the tag. A tag key can have a maximum of 512 characters and cannot be empty. Key
+	// must begin with a letter, end with a letter, number or underscore, and must contain only alphanumeric
+	// characters and the following special characters `_ . -`.
+	//
```
Done

fahlmant commented on api/hypershift/v1beta1/azure.go:
```diff
@@ -735,3 +780,24 @@ type AzureAuthenticationConfiguration struct {
+	// +kubebuilder:validation:Pattern=`^[a-zA-Z]([0-9A-Za-z_.-]*[0-9A-Za-z_])?$`
+	Key string `json:"key"`
+	// value is the value part of the tag. A tag value can have a maximum of 256 characters and cannot be empty. Value
+	// must contain only alphanumeric characters and the following special characters `_ + , - . / : ; < = > ? @`.
```
Done

bryan-cox commented:
> Please convert this to a draft this to a draft PR until its ready for review so it doesn't trigger CI jobs on every push.

@fahlmant can you please convert this to draft when you have a moment?

fahlmant commented:
@bryan-cox it should be ready for review. I'm not sure why the e2e is failing as didn't touch AzureWorkloadIdentities

bryan-cox commented on api/hypershift/v1beta1/azure.go:
```diff
@@ -98,6 +98,26 @@ type AzureNodePoolPlatform struct {
+	// +listType=map
+	// +listMapKey=key
+	// +optional
+	ResourceTags []AzureResourceTag `json:"resourceTags,omitempty"`
```
https://github.com/openshift/hypershift/pull/6846/files#r2364174798 still needs to be addressed

bryan-cox reviewed (commented):
Hey @fahlmant - thanks for the PR! Just a few things:
- Please follow the github pull request template and update the PR description
- Please identify which Jira ticket this is solving
- There seems to be one comment that wasn't addressed

Otherwise seems good to me.

bryan-cox commented:
/close

Please feel free to reopen if this is still needed.
